python/pipeline/stack.py

```diff
@@ -977,6 +978,129 @@ def make(self, key):
+        #  - Windows where the final median intensity is below 10 are removed
+        #      - Attempts to replace this with a percentile all fail
+        #  - This code drops guessed depths > 95th-percentile and < 5th-percentile to be more robust to outliers
+
+        def surface_eqn(data, a, b, c, d, f):
+            x, y = data
+            return a * x ** 2 + b * y ** 2 + c * x + d * y + f
+
+        # SETTINGS
+        # Note: Intial parameters for fitting set further down
+        r = 50  # Radius of square in pixels
+        upper_threshold_percent = 0.6  # Surface median intensity should be in the bottom X% of the *range* of medians
+        gaussian_blur_size = 5  # Size of gaussian blur applied to slice
+        min_points_allowed = 10  # If there are less than X points after filtering, throw an error
+        bounds = (
+        [0, 0, np.NINF, np.NINF, np.NINF], [np.Inf, np.Inf, np.Inf, np.Inf, np.Inf])  # Bounds for paraboloid fit
+        ss_percent = 0.40  # Percentage of points to subsample for robustness check
+        num_iterations = 1000  # Number of iterations to use for robustness check
+
+        print('Calculating surface of brain for stack', key)
+        full_stack = (stack.PreprocessedStack & key).fetch1('resized')
+        depth = full_stack.shape[0]
+        height = full_stack.shape[1]
+        width = full_stack.shape[2]
+
+        surface_guess_map = []
+        r_xs = np.arange(r, width - width % r, r * 2)[1:-1]
+        r_ys = np.arange(r, height - height % r, r * 2)[1:-1]
+        full_mesh_x, full_mesh_y = np.meshgrid(np.arange(width), np.arange(height))
+
+        # Surface z should be below this value
+        z_lim = int(depth / 2)
+        # Mean intensity of the first frame in the slice should be less than this value
+        z_0_upper_threshold = np.percentile(full_stack, 30)
+
+        for x in r_xs:
+            for y in r_ys:
+                stack_slice_medians = np.percentile(full_stack[0:z_lim, y - r:y + r, x - r:x + r], 50, axis=(1, 2))
+                blurred_slice = ndimage.gaussian_filter1d(stack_slice_medians, gaussian_blur_size)
+
+                upper_threshold_value = upper_threshold_percent * (
+                            (blurred_slice.max() - blurred_slice.min()) - blurred_slice.min())
+                upper_threshold_idx = np.where(blurred_slice > upper_threshold_value)[0][0]
+
+                stack_slice_derivative = ndimage.sobel(blurred_slice)
+                surface_z = np.argmax(stack_slice_derivative)
+
+                if ((surface_z < upper_threshold_idx) and (blurred_slice[0] < z_0_upper_threshold) and
+                        (blurred_slice[-1] > 10)):
+                    surface_guess_map.append((surface_z, y, x))
+
+        if len(surface_guess_map) < min_points_allowed:
+            raise Exception(f"Surface calculation could not find enough valid points for {key}. Only {len(surface_guess_map)} detected")
```

Comment:
you could use PipelineException (it's already imported in stack), it's kind of the generic exception we use for pipeline errors. Advantage is that at least we know for sure that we are the ones raising that exception rather than Exception which could be raised by numpy or by some other package.